### ISSUE TITLE
Why skip last batch?

### ISSUE THREAD

**Issue by Tiamo666** [layumi/Person-reID-triplet-loss#2]
Hi, Zedong, Thanks a lot for your work.
I was a little confused about the following code:
       if now_batch_size<opt.batchsize: # next epoch
                     continue
Does it will influence the performance?

**Comment by layumi** [comment]
Hi @Tiamo666 
Generally, it may not compromise the final performance. But in some cases, it does.
1. For example, after you have trained most images in the training set, you may left one image.
Using batch normalisation, one image will output zero and abnormal prediction. 
2. Small batchsize also will affect the running mean and std in the batchnorm layer.

**Comment by layumi** [comment]
@Tiamo666 
Sure. You can sample another batch as negative pool.
But it may be more efficient by directly using other `positive` data as negative sample.

**Comment by Tiamo666** [comment]
@layumi  OK. And is it necessary to do the random permutation of "nf_data"(in train_new.py line ~217)? Cause your default opt.poolsize is 128(equal to the batchsize of nf_data), and after sorting score, the result will be the same.

**Comment by layumi** [comment]
@Tiamo666 
It is designed for the small poolsize. 
If you using the biggest pool size, it is not neccessary.
